tools/eslint-rules/rules/utils/typescript-ast-utils.ts

```diff
@@ -0,0 +1,254 @@
+    return false;
+  }
+
+  const expr = stmt.expression;
+  if (!ts.isBinaryExpression(expr)) {
+    return false;
+  }
+
+  if (expr.operatorToken.kind !== ts.SyntaxKind.EqualsToken) {
+    return false;
+  }
+
+  if (!ts.isPropertyAccessExpression(expr.left)) {
+    return false;
+  }
+
+  if (!ts.isIdentifier(expr.left.name)) {
+    return false;
+  }
+
+  return expr.left.name.text === propertyName;
+}
+
+/**
+ * Checks if a type implements a specific interface directly or through inheritance
+ * @param type - The TypeScript type to check
+ * @param interfaceName - The name of the interface to check for
+ * @param checker - TypeScript type checker instance
+ * @returns True if the type implements the interface
+ */
+export function implementsInterface(
+  type: ts.Type,
+  interfaceName: string,
+  checker: ts.TypeChecker
+): boolean {
+  if (!type.isClass()) {
+    return false;
+  }
+
+  const symbol = type.getSymbol();
+  if (!symbol) {
+    return false;
+  }
+
+  // Check if the type directly implements interface through properties
+  const interfaces = type
+    .getProperties()
+    .map((prop) => prop.declarations?.[0]?.parent)
+    .filter((parent): parent is ts.Node => parent !== undefined);
+
+  if (interfaces.some((iface) => isInterfaceWithName(iface, interfaceName))) {
+    return true;
+  }
+
+  // Check if any interface in the heritage chain matches
+  if (symbol.declarations?.[0]) {
+    const declaration = symbol.declarations[0];
+    if (ts.isClassDeclaration(declaration) && declaration.heritageClauses) {
+      const implementsInterface = declaration.heritageClauses.some((clause) => {
+        if (clause.token !== ts.SyntaxKind.ImplementsKeyword) {
+          return false;
+        }
+        return clause.types.some((type) => {
+          const typeSymbol = checker.getSymbolAtLocation(type.expression);
+          return typeSymbol?.getName() === interfaceName;
+        });
+      });
+
+      if (implementsInterface) {
+        return true;
+      }
+    }
+  }
+
+  // Check base class recursively
+  const baseTypes = type.getBaseTypes();
+  if (!baseTypes) {
+    return false;
+  }
+
+  return baseTypes.some((baseType) =>
+    implementsInterface(baseType, interfaceName, checker)
+  );
+}
+
+/**
+ * Checks if a property is definitely initialized in the class or its parent classes
+ * @param classType - The TypeScript type representing the class
+ * @param propertyName - The name of the property to check
+ * @returns True if the property is definitely initialized
+ */
+export function isPropertyInitialized(
+  classType: ts.Type,
+  propertyName: string
+): boolean {
+  const symbol = classType.getSymbol();
+  if (!symbol?.declarations?.[0]) {
+    // Check parent class if no declarations in current class
+    const baseTypes = classType.getBaseTypes();
+    if (!baseTypes) {
+      return false;
+    }
+    return baseTypes.some((baseType) =>
+      isPropertyInitialized(baseType, propertyName)
+    );
+  }
```

Comment:
What doest it mean that a class symbol doesn't have a "declaration"? And why in such case we should check the base type?

I understood that in line 245 we start fallbacking to check ancestor classes chain.

Reply:
Answering the first question, based on my knowledge it can happen in the real case: Syntetic types, union types, anonymous classes, classes from declaration files and. But indeed, we don;t need to rely on that in our case, but we just need to satisfy typing.

We could pass node and adjust the function to do all checks based on it, but to check ancestors, we would have to involve symbol either way.

That said, I'd still use Symbol, but i simplified the function's logic

Reply:
Super readable now. Many thanks!